feat: implement stars and planet display position logic

mehdi-hamirifou opened the pull request:
This issue resolves issue [#181](https://github.com/LookUpGroup27/LookUp/issues/181)

### Features:

**Planets' Positions Logic:**

- Added logic for calculating and updating the positions of planets based on their celestial coordinates and time-based computations. (Commit: 3d59fa0)

**Celestial Object Utilities:**

- Implemented CelestialObjectUtils to centralize helper functions for position computations of celestial objects, ensuring modular and reusable code. (Commit: 9430c83)
Star Data Handling:

- Updated star data handling classes to incorporate location and time-based computations for accurate positioning in the skymap. (Commit: 9bfc003)
Refactor:

### Stars Package Reorganization:
- Moved the stars package into the map package to streamline the project structure and improve maintainability. (Commit: 0541667)

### Testing:

**New Test Implementations:**

- PlanetsRepositoryTest to validate the newly introduced planets' position logic. (Commit: ce08ce9)
- CelestialObjectUtilsTest to ensure the correctness of celestial computations. (Commit: 16aabe5)

**Updated Tests:**

Updated existing tests to align with the refactored and extended logic:
- StarsLoaderTest for loading star data. (Commit: 05b5a45)
- StarRepositoryTest to match the updated star data handling. (Commit: 55d7338)


**Note:** _Due to the current uncertainty surrounding the OpenGL implementation, which is affected by several bugs, it is not yet possible to verify the accuracy of the display. As a result, this pull request has been submitted as a draft._

**Post-Note Changes:** After fixing the OpenGL rendering issues and adjusting the relevant classes for display testing, it became evident that inaccuracies persisted due to improper unit conversions. Additional commits were therefore introduced to resolve these conversion errors, ensuring the display logic now functions correctly and results in a sky map that accurately reflects real-world celestial conditions.

### Documentation: 

The features implemented in this pull request involved complex astronomical concepts. To provide better insight into the underlying computations and the tools utilised, here are the main documentation resources that informed this work:

- [HYG Database (Astronexus)](https://www.astronexus.com/hyg)
- [JPL Horizons API Documentation](https://ssd-api.jpl.nasa.gov/doc/horizons.html)
- [Celestial Horizon Coordinates Calculator](http://www.convertalot.com/celestial_horizon_co-ordinates_calculator.html)
- [Alt-Az Conversion (astrogreg.com)](https://astrogreg.com/convert_ra_dec_to_alt_az.html)
- [Stellarium Web](https://stellarium-web.org/)
- [U.S. Naval Observatory FAQs - GAST](https://aa.usno.navy.mil/faq/GAST)
- [MATLAB Central: RA/Dec Alt/Az Conversion Explorer](https://ch.mathworks.com/matlabcentral/fileexchange/134556-ra-dec-alt-az-conversion-explorer)
- [Stellarium Remote Control API Docs](https://stellarium.org/doc/23.0/remoteControlApi.html)
- [French Wikipedia: Equatorial Coordinate System](https://fr.wikipedia.org/wiki/Syst%C3%A8me_de_coordonn%C3%A9es_%C3%A9quatoriales)
- [Wikipedia: Horizontal Coordinate System](https://en.wikipedia.org/wiki/Horizontal_coordinate_system)

mehdi-hamirifou committed:
feat: Implement Celestial Object Utils to group helper functions for celestial objects' positions' computations

mehdi-hamirifou committed:
feat: Update star data handling classes to handle location and time based computations

- Added fields to StarData data class to hold necessary data for new computations
- Refactored getStars to loadStarsFromCSV in StarDataRepository to match changes in StarData
- Added updateStarPositions() to StarDataRepository to allow for live updates of positions according to time and location
-Removed unnecessary calculateSizeFromMagnitude function
-  Updated StarsLoader to match changes in StarData and StarDataRepository

mehdi-hamirifou committed:
feat: Comment out calls to star data related classes in MapRenderer

- Comment out parts of the code in MapRenderer to postpone until later the implementation of its final version with new data handling

mehdi-hamirifou committed:
test: Implement test class for CelestialObjectUtils

mehdi-hamirifou committed:
test: Update StarRepositoryTest class to match new implementation

mehdi-hamirifou committed:
test: Update StarsLoaderTest class to match new implementation

mehdi-hamirifou committed:
feat: Implement planets' positions logic

- Created PlanetData data class to represent planets
- Implemented PlanetsRepository to retrieve and hold planets' data and handle it

mehdi-hamirifou committed:
refactor: Move stars' package to the map package for clearer structure

mehdi-hamirifou committed:
test: Implement PlanetsRepositoryTest class

- Added tests for PlanetsRepository
- Added gradle dependency for mocking web servers

mehdi-hamirifou committed:
test: Remove debugging messages from convertToHorizonCoordinates function

mehdi-hamirifou committed:
style: Apply ktfmtFormat

mehdi-hamirifou committed:
fix: Remove unused variable in StarDataRepository

mehdi-hamirifou committed:
Merge branch 'main' into feat-implement-stars-and-planet-display-position-logic-

mehdi-hamirifou committed:
fix: Re comment out code in MapRenderer after solving conflicts

mehdi-hamirifou committed:
style: Apply ktfmtFormat

mehdi-hamirifou committed:
feat: Update MapRenderer to display stars

mehdi-hamirifou committed:
fix: Add gradle dependencies needed for connected checks

mehdi-hamirifou committed:
feat: Correct CelestialObjectUtils functions

- Changed computations
- Refactored methods to improve readability

mehdi-hamirifou committed:
test: Update CelestialObjectUtilsTest to match changes

mehdi-hamirifou committed:
feat: Correct PlanetData fetching and handling

- Added Texture Id field to PlanetData
- Updated PlanetRepository methods to ensure correctness
- Added comments and documentation

mehdi-hamirifou committed:
test: Update PlanetsRepositoryTest to match changes

mehdi-hamirifou committed:
feat: Improve starData handling

- Removed unnecessary spectralClass Parameter in StarData
- Refactored StarDataRepository functions to improve efficiency and practicality
-  Refacted StarsLoader to match changes
- Improved Documentation

mehdi-hamirifou committed:
test: Updated StarDataRepositoryTest to match changes

mehdi-hamirifou committed:
feat: Update renderables for good display

mehdi-hamirifou committed:
feat: Update sky map MVVM to ensure good integration of location and time for stars display

- Made MapScreen display only if location is provided to ensure display correctness
- Refactored MapViewModel to instantiate repositories and pass necessary location and context parameters to MapRenderer at creation
- Updated MapRenderer to integrate location and star data handling for the new display logic
- Removed call to updateContext function in MapSurfaceView to match changes
- Updated MainActivity to match changes

mehdi-hamirifou committed:
test: Update sky map related test to match changes

mehdi-hamirifou committed:
Merge branch 'main' into feat-implement-stars-and-planet-display-position-logic-

mehdi-hamirifou committed:
style: Apply ktfmtFormat

mehdi-hamirifou committed:
feat: Add loading map Text to prevent ambiguity on map screen display

mehdi-hamirifou committed:
style: Apply ktfmtFormat

raniabbrk reviewed (changes requested):
Great job on this  PR! It is very impressive! You’ve tackled a crucial part of the project, and the display position logic for the stars and planets looks great. I also wanted to mention that your implementation is very clean and well-organized. It’s clear you put thought into making the code reusable and scalable, great work on that! 

That being said, I noticed a couple of things:

---

### **1. Moon in the Positioning Logic**

Right now the Moon’s position and scale are hardcoded in its class since we didn't have the display position logic at the time. Since you’ve already implemented the logic for celestial objects, it would be nice to extend that to include the Moon so everything is clean when we merge your PR. I did some research and the Moon should be included in NASA's JPL Horizons API. Here are my ideas on how you can maybe include the moon in the display position logic:

First, in the `planets` list, include the Moon with its Horizons ID (`301`) but without a fixed texture ID:

```kotlin
planets.add(
    PlanetData("Moon", "301", textureId = null) // Texture will be updated dynamically
)
```

Then, modify the `updatePlanetsData()` method in `PlanetsRepository` to handle the Moon just like the planets. 

Then we should dynamically assign the Moon’s texture based on its phase. Maybe it could work if we do it like the following:
```kotlin
planets.forEach { planet ->
    val position = fetchPlanetPosition(latitude, longitude, planet.id)
    if (position != null) {
        planet.ra = position.first
        planet.dec = position.second

        val (azimuth, altitude) = CelestialObjectsUtils.convertToHorizonCoordinates(
            ra = planet.ra,
            dec = planet.dec,
            latitude = latitude,
            localSiderealTime = siderealTime
        )

        planet.cartesian = CelestialObjectsUtils.convertToCartesian(azimuth, altitude)

        // Dynamically update the Moon’s texture
        if (planet.name == "Moon") {
            planet.textureId = Moon.getCurrentMoonPhaseTextureId()
        }
    }
}
```

Then you'll need to modify the `Moon` class to use the dynamically calculated position and texture from `PlanetData`:

```kotlin
class Moon(
    context: Context,
    planetData: PlanetData
) : Planet(
    context = context,
    name = planetData.name,
    position = floatArrayOf(
        planetData.cartesian.first,
        planetData.cartesian.second,
        planetData.cartesian.third
    ),
    textureId = planetData.textureId ?: R.drawable.default_moon_texture // Fallback texture
)
```
I hope it’s doable to include the Moon in this PR using the steps I outlined. I’m not entirely sure if my suggestions will work as is, but I hope they provide a good starting point. Let me know if you have any questions or run into issues. If it turns out that it’s not feasible to include the Moon here, we can figure out an alternative solution in another PR :).

---

### **2. Move `getJulianDay` to `CelestialObjectUtils`**

The `getJulianDay` method is currently in the Moon class, but it’s a general utility that isn’t Moon-specific. I know you weren't in charge of this class, but since you created  `CelestialObjectUtils` class , I think it would be nice if you could move `getJulianDay` to it. It would make more sense and ensure the code is clean! (You would have to  replace the `getJulianDay` logic in the Moon class with a call to `CelestialObjectUtils.getJulianDay`).

---

Also, just a quick note: When I implemented the 'Planet' class in previous sprints, I added a 'setScale' function because I thought it might come in handy when we implemented the display position logic. Since you didn’t end up using it in this PR, feel free to delete it if you think it’s unnecessary.

---

Overall, this task was very complex and hard to implement and you did an amazing job at it!

mehdi-hamirifou committed:
refactor: Move getJulianDay function to CelestialObjectsUtils

mehdi-hamirifou committed:
feat: Update PlanetsRepository to include Moon's display position retrieval and update logic

mehdi-hamirifou committed:
refactor: Delete old moon render logic from MapRenderer

mehdi-hamirifou committed:
refactor: Delete unused setScale function from Planet class

mehdi-hamirifou committed:
fix: Ensure Map Screen refreshes when location permission is given

- Added refresh button in case location wasn't provided at the first loading of the map

raniabbrk reviewed (approved):
It looks good to me now. Ready to merge.

mehdi-hamirifou committed:
Merge branch 'main' into feat-implement-stars-and-planet-display-position-logic-

mehdi-hamirifou committed:
feat: Assign new textures to planets in PlanetsRepository

mehdi-hamirifou committed:
refactor: Update StarsLoader's mapping to match new constructor of Star class

mehdi-hamirifou committed:
style: Apply ktfmtFormat
